Various fixes and improves

builtinnya opened the pull request:
I've done some fixes and improves.

Important changes are:
- sleep before starting Sevabot to avoid skype.Attach() connection refuse
- easily customizable message handler framework

builtinnya committed:
Added a sleep before starting Sevabot to avoid skype.Attach() failing

Other minor fixes
 - Added variables for the username and X server display number
 - Added some indents
 - Similar fixes on `start-vnc.sh'

builtinnya committed:
Added `LOG_LEVEL' variable for setting

`LOG_LEVEL' can be either 'INFO' or 'DEBUG'.
This variable overrides `--verbose' option.

builtinnya committed:
Added a customizable message handler framework

Now you can easily add built-in commands and other advanced
handlers by defining handler classes.
Also added a built-in command for calling.

Other minor fixes
 - Stopped using `shlex' to parse commands because
   msg.Chat.Body.encode('utf-8') seems to fail on some languages
   e.g. Japanese.

builtinnya committed:
Added a fancy dice roll module

miohtama commented:
Looks good.

Something needed to get this forward
- Can you update the documentation how one should add and include message handlers? Maybe if you can first explain here briefly and let's see how we can up an code example

miohtama commented on sevabot/frontend/main.py:
```diff
@@ -70,7 +70,15 @@ def main(settings="settings.py", verbose=False, daemon=False):
     # Config logging
-    level = logging.DEBUG if verbose else logging.INFO
+    if getattr(settings, "LOG_LEVEL", None):
+        if settings.LOG_LEVEL == "INFO":
```
I think here we could do simpler

```
 level = getattr(logging, settings.LOG_LEVEL, "INFO")  # Automatically maps text string to the logging level, defaults to INFO
```

builtinnya commented:
Your simpler version looks nice.

Here is an explanation how to add and include message handlers:

If you want to add a completely new handler, you define
- a class whose name ends with `Handler` (e.g. `CommandHandler`) in `sevabot.bot.handlers` module

Once defined, the handler classes will be automatically instantiated and initialized when `Sevabot.start()`.
In handler classes, you must define two methods:
- `init(self, skype)` that is used to initialize a handler object
- `handle(self, msg, status)` that is automatically called on every message received and sent

If you want to add just a built-in command (`!command`), you define in `CommandHandler` class
- a method whose name starts with `builtin_`  (e.g. `builtin_reload(self, args, msg, status)` )

Once defined, the method will be automatically called if `!method_name_after_builtin_` message received
(e.g. `!reload`.)

builtinnya commented:
I wrote in `70ba7a8` that `msg.Chat.Body.encode('utf-8') seems to fail on some languages (e.g. Japanese)` 
but that was wrong.

It is the undecoded (remaining encoded) string that causes the error when calling `keyword.startswith('!')`.
We can fix it by calling `decode('utf-8')` on each word before checking `keyword.startswith('!')`.

builtinnya committed:
Made getting LOG_LEVEL setting simpler

builtinnya committed:
Made sure that split words are decoded before use

Now use `shlex` to split a message into words.

builtinnya committed:
Added base class for message handlers

All message handlers must inherit `HandlerBase`.
Users should put custom handlers in `custom_handlers.py`.

miohtama commented:
Working on this now

miohtama commented:
I think we stil need to update manual regarding new build-in commands (call). I am not sure what one is supposed to do with that.

miohtama commented:
Also moved to work to dev branch...  I got inspired by your ideas and I am working to make stateful, reloadable, Python handlers

builtinnya commented:
Great!
Stateful handlers are what I want.

The built-in `!call` command is for hosting a conference call of a chat so that the caller (one typed !call) can feel
free to disconnect from the call.
This is just my local need. Now I think `!call` should not be a built-in command.

miohtama commented:
Now when we have stateful handlers maybe we can refactor !call to be its own handler? 

I think we just need to route more Skype events to the handlers - now it only has callback for chat message.

After this is done we could release version 1.2 :)

builtinnya commented:
I'm working on the stateful handler version of the `!call`.

I think callback manager (adder/remover) is enough for the handlers to handle every Skype event.
Each added callback will be called in order.
